app/vmalert/manager.go

```diff
@@ -156,11 +156,14 @@ func (m *manager) update(ctx context.Context, groupsCfg []config.Group, restore
 		var wg sync.WaitGroup
 		for _, item := range toUpdate {
 			wg.Add(1)
+			// cache the cancel func before updating the Group, as it might change during the Update
+			evalCancel := item.old.EvalCancelFn
 			go func(old *rule.Group, new *rule.Group) {
 				old.UpdateWith(new)
 				wg.Done()
 			}(item.old, item.new)
-			item.old.InterruptEval()
```

Comment:
quick Q: why not just move this line before `go func(old *rule.Group, new *rule.Group)`

Reply:
The rules evaluation and group update are done within the same goroutine. Group can't be updated while it evaluates rules - the update will have to wait for its turn.
So when we want to update Group, we want:
1. To start waiting for our turn
2. To cancel ongoing evaluation of rules, so we don't wait for them to finish.

If we do 2 first, then it is possible that rules execution will start before the update... oh. And this evaluation will immediately finish because all evals will return `context canceled` err. You're right!

See update